src/scope.ml

```diff
@@ -118,6 +118,95 @@ type mode =
       order variables that will be bound in the representation of the RHS. *)
 
+(* splitFunArgs t decomposes the parser term [t] (which is expected to be an application),
+  into the function symbol and a list of its actual arguments. It is implemented 
+  as a tail-recursive function. *)
+let splitFunArgs : p_term -> (p_term * (p_term list)) = fun t ->
+  let rec splitFunArgs_aux : p_term -> (p_term list) -> (p_term * (p_term list)) = fun t args ->
+    match t.elt with
+    | P_Appl(u,v) -> splitFunArgs_aux u (v::args)
+    | _           -> (t, args)
+  in
+  splitFunArgs_aux t []
+
+(* Just for debugging *)
+let boolList_to_string : bool list -> string = fun l ->
+  let rec boolList_to_string_aux : bool list -> string = fun l ->
+  match l with
+  | [] -> "] \n"
+  | h::[] -> (string_of_bool h) ^"] \n"
+  | h::t -> (string_of_bool h) ^ "; " ^ (boolList_to_string_aux t)
+  in "[" ^ boolList_to_string_aux l
+
+(** [getParamsImplicitness ss t] returns a list representing the formal parameters of 
+    a parser term [t] *)
+let getParamsImplicitness : sig_state -> p_term -> bool list = fun ss t ->
+  match t.elt with
+  | P_Prod(args, _) -> List.map (fun x -> match x with (_,_,i) -> i) args
+  | P_Abst(args,_)  -> 
+      begin
+        (* print_string "debug : P_Abst found! \n"; *)
+        List.map (fun x -> match x with (_,_,i) -> i) args
+      end
+  | P_Vari(id, imp) -> if imp=FullyExplicit then [] 
+    else
+    begin
+      (* print_string "debug : P_Vari found! \n"; *)
+      match (StrMap.find_opt (snd id.elt) ss.in_scope) with
+      | Some(res) -> let implicits = (fst res).sym_implicits in
+                      begin
+                        (* print_string "debug : found it in in_scope ! \n"; *)
+                        (* print_string (boolList_to_string implicits); *)
+                        implicits
+                      end
+      | None      -> 
+        (match (StrMap.find_opt (snd id.elt) ss.builtins) with
+        | Some(res) -> (* print_string "debug : found it in builtins ! \n"; *) 
+                       (fst res).sym_implicits
+        | None -> (* print_string "debug : not found ! \n"; *)
+                     [])
+    end
+  | P_Patt(_,_)     -> (* print_string "debug : Pattern ! \n"; *)
+                       [] (* TO DO : see if we want to have implicits for patterns as well, but I don't think so *)
+  | _               -> []
+
+(** [addImplicitArgs args param] builds the full arguments list, using the given arguments 
+    [args] and with the insertion of "_" for the implicit arguments, following the information 
+    provided by the formal parameters [param] *)
+let addImplicitArgs : bool list -> p_term list -> p_term list = fun param args ->
+  let rec addImplicitArgs_aux : bool list -> p_term list -> p_term list -> p_term list = 
+    fun param args fullArgs ->
+    match (param,args) with
+    | (isImplicit::ps, arg1::args') -> 
+      if isImplicit then
```

Comment:
Since isImplicit is a boolean, you can directly do pattern matching on true/false instead of using if:
    match (param,args) with
    | (true::ps, arg1::args') -> ...

Reply:
I find it clearer with 
    | (isImplicit::ps, arg1::args') -> 
      if isImplicit then
as it tells us what this boolean is by naming it.

Reply:
I agree with Frédéric. You can always add a comment in each case to explain what is the situation.

Reply:
Ok.